interfaces/builtin/avahi_control_test.go

```diff
@@ -0,0 +1,171 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin_test
+
+import (
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/builtin"
+	"github.com/snapcore/snapd/interfaces/dbus"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/testutil"
+)
+
+type AvahiControlInterfaceSuite struct {
+	iface    interfaces.Interface
+	plug     *interfaces.Plug
+	appSlot  *interfaces.Slot
+	coreSlot *interfaces.Slot
+}
+
+var _ = Suite(&AvahiControlInterfaceSuite{
+	iface: builtin.MustInterface("avahi-control"),
+})
+
+const avahiControlConsumerYaml = `name: consumer
+apps:
+ app:
+  plugs: [avahi-control]
+`
+
+const avahiControlProducerYaml = `name: producer
+apps:
+ app:
+  slots: [avahi-control]
+`
+
+const avahiControlCoreYaml = `name: core
+slots:
+  avahi-control:
+`
+
+func (s *AvahiControlInterfaceSuite) SetUpTest(c *C) {
+	s.plug = MockPlug(c, avahiControlConsumerYaml, nil, "avahi-control")
+	s.appSlot = MockSlot(c, avahiControlProducerYaml, nil, "avahi-control")
+	s.coreSlot = MockSlot(c, avahiControlCoreYaml, nil, "avahi-control")
+}
+
+func (s *AvahiControlInterfaceSuite) TestName(c *C) {
+	c.Assert(s.iface.Name(), Equals, "avahi-control")
+}
+
+func (s *AvahiControlInterfaceSuite) TestSanitizeSlot(c *C) {
+	c.Assert(s.coreSlot.Sanitize(s.iface), IsNil)
+	c.Assert(s.appSlot.Sanitize(s.iface), IsNil)
+	// avahi-control slot can now be used on snap other than core.
+	slot := &interfaces.Slot{SlotInfo: &snap.SlotInfo{
+		Snap:      &snap.Info{SuggestedName: "some-snap"},
+		Name:      "avahi-control",
+		Interface: "avahi-control",
+	}}
+	c.Assert(slot.Sanitize(s.iface), IsNil)
+}
+
+func (s *AvahiControlInterfaceSuite) TestSanitizePlug(c *C) {
+	c.Assert(s.plug.Sanitize(s.iface), IsNil)
+}
+
+func (s *AvahiControlInterfaceSuite) TestAppArmorSpec(c *C) {
+	// on a core system with avahi slot coming from a regular app snap.
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	// connected plug to app slot
+	spec := &apparmor.Specification{}
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, nil, s.appSlot, nil), IsNil)
+	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "name=org.freedesktop.Avahi")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, `peer=(label="snap.producer.app"),`)
```

Comment:
It might be nice to verify a specific rule from avahi-observe and from avahi-control for AddConnectedPlug and AddConnectedSlot since avahi-control pulls in both.